File: code/analysis-code/functions/virus-transform-function.R
# -----------------------------------------------------------------------------
# virus-transform-function.R
# -----------------------------------------------------------------------------
# PURPOSE
#   Centralize virus load transformations so data, model predictions, and plots
#   stay consistent. Adjust the transform in ONE place below.
# -----------------------------------------------------------------------------

# Name used for virus quantity throughout the codebase.
virus_quantity_name <- "VirusLoad"

# Transformation mode for virus values. Options:
#   - "identity": no transform (keep in natural units)
#   - "log10_p1": log10(V + 1)
#   - "log10_clamp1": log10(max(V, 1))
virus_transform_mode <- "log10_p1"

# Apply the selected virus transform.
transform_virus <- function(x, mode = virus_transform_mode) {
  x <- as.numeric(x)

  switch(
    mode,
    identity = x,
    log10_p1 = log10(pmax(x, 0) + 1),
    log10_clamp1 = log10(pmax(x, 1)),
    stop("Unknown virus_transform_mode: ", mode)
  )
}

# Invert the selected virus transform (for plotting raw values).
inverse_transform_virus <- function(x, mode = virus_transform_mode) {
  x <- as.numeric(x)

  switch(
    mode,
    identity = x,
    log10_p1 = pmax(0, 10^x - 1),
    log10_clamp1 = 10^x,
    stop("Unknown virus_transform_mode: ", mode)
  )
}

File: code/julia/src/virus-transform-function.jl
"""
Virus transform helpers
=======================

The R workflow transforms viral load before fitting. Keeping the transform in
one location avoids subtle mismatches across fitting, diagnostics, and plotting.

Current default
---------------
`log10_p1`: `log10(max(x, 0) + 1)`
"""

const virus_quantity_name = "VirusLoad"
const virus_transform_mode = "log10_p1"

"""
Apply the configured virus transform to a scalar.
"""
function transform_virus(x::Real; mode::String = virus_transform_mode)
    if mode == "identity"
        return x
    elseif mode == "log10_p1"
        # Keep numeric type generic (e.g., ForwardDiff.Dual in stiff ODE solvers).
        return log10(max(x, zero(x)) + one(x))
    elseif mode == "log10_clamp1"
        return log10(max(x, one(x)))
    else
        error("Unknown virus transform mode: $mode")
    end
end

"""
Vector overload for convenience and readability.
"""
function transform_virus(x::AbstractVector{<:Real}; mode::String = virus_transform_mode)
    return [transform_virus(xi; mode = mode) for xi in x]
end

"""
Inverse virus transform.

Useful for plotting trajectories on the natural scale if needed.
"""
function inverse_transform_virus(x::Real; mode::String = virus_transform_mode)
    if mode == "identity"
        return x
    elseif mode == "log10_p1"
        return max(zero(x), (10.0^x) - one(x))
    elseif mode == "log10_clamp1"
        return 10.0^x
    else
        error("Unknown virus transform mode: $mode")
    end
end
